File: C# part 2/Arrays/11.BinarySearch/BinarySearch.cs
/*
 Problem 11. Binary search

    Write a program that finds the index of given element in a sorted array of integers by using the Binary search algorithm.

 */
namespace _11.BinarySearch
{
    using System;
    class BinarySearch
    {
        static void Main()
        {

            Console.Write("Enter array length: ");
            int length = int.Parse(Console.ReadLine());

            int[] numbers = new int[length];

            Console.WriteLine("Enter integer values");
            for (int i = 0; i < numbers.Length; i++)
            {
                Console.Write("Array[{0}] = ", i);
                string input = Console.ReadLine();
                int number;
                bool isNumber = int.TryParse(input, out number);
                if (isNumber)
                {
                    numbers[i] = number;
                }
                else
                {
                    Console.WriteLine("Input string was not in a correct format");
                }
            }

            Console.Write("Enter element: ");
            int element = int.Parse(Console.ReadLine());

            int minIndex = 0;
            int maxIndex = numbers.Length - 1;
            int position = 0;
            Array.Sort(numbers);
            while (minIndex < maxIndex)
            {
                int middIndex = (minIndex + maxIndex) / 2;

                if (numbers[middIndex] == element)
                {
                    position = middIndex;
                    break;
                }
                else if (numbers[middIndex] < element)
                {
                    minIndex = middIndex + 1;
                }
                else
                {
                    maxIndex = middIndex - 1;
                }
            }

            Console.WriteLine("Sorted Array");
            for (int i = 0; i < numbers.Length; i++)
            {
                Console.WriteLine("Array[{0}] = {1}",i, numbers[i]);
            }
            Console.WriteLine("Position: {0}", position);
        }
    }
}

File: JavaScript Fundamentals/Javascript Arrays/scripts/binarySearch.js
jsConsole.writeLine('--------------------------------------------------------------------');
jsConsole.writeLine('Problem 7: Write a script that finds the index of given element in a sorted array of integers by using the binary search algorithm.')
jsConsole.writeLine('--------------------------------------------------------------------');
console.log('--------------------------------------------------------------------');
console.log('Problem 7: Write a script that finds the index of given element in a sorted array of integers by using the binary search algorithm.')
console.log('--------------------------------------------------------------------');

console.log('Before sort: [9, 7, 77, 5, 5, 3, 2, 1, -4, 6, 8, 10]');
jsConsole.writeLine('Before sort: [9, 7, 77, 5, 5, 3, 2, 1, -4, 6, 8, 10]');


var numbers = [9, 7, 77, 5, 5, 3, 2, 1, -4, 6, 8, 10];
function orderBy(a, b) {
    return (a == b) ? 0 : (a > b) ? 1 : -1
};
numbers.sort(orderBy);
jsConsole.writeLine('After sort: [' + numbers.join(", ") + ']');

var value = 9;
var left = 0;
var right = numbers.length - 1;

while (left <= right) {

    var middle = ((left + right) / 2);
    middle = parseInt(middle);
    if (numbers[middle] == value) {
        break;
    }
    else if (numbers[middle] > value) {
        right = middle - 1;
    }
    else {
        left = middle + 1;
    }
}

jsConsole.writeLine("Number " + value + " at index " + middle);
console.log("Number " + value + " at index " + middle);